File: Students/YeuksingChan/lc252.js
// Problem Statement #

// Given an array of intervals representing ‘N’ appointments, find out if a person can attend all the appointments.
//*******************
//For this problem, give two solutions, one with a for loop and one with a while loop
//*******************
class Interval {
  constructor(start, end) {
    this.start = start;
    this.end = end;
  }

  print_interval() {
    process.stdout.write(`[${this.start}, ${this.end}]`);
  }
}

const can_attend_all_appointments = (arr) => {
  arr.sort((a, b) => a.start - b.start);
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[i].end > arr[j].start) {
        return false;
      }
    }
  }
  return true;
};

const can_attend_all_appointments2 = (arr) => {
  arr.sort((a, b) => a.start - b.start);
  let i = 0;
  while (i < arr.length - 1) {
    let j = i + 1;
    while (j < arr.length) {
      if (arr[i].end > arr[j].start) {
        return false;
      }
      j++;
    }
    i++;
  }
  return true;
};

console.log(
  `Can attend all appointments: ${can_attend_all_appointments2([
    new Interval(1, 4),
    new Interval(2, 5),
    new Interval(7, 9),
  ])}`
);
// [1,4] [2,5] [7,9]  arrLength: 3, lastInterval: 2, currentIndex : 0
// [1,4] [2,5] [7,9]  4 ?> 2 yes -> FALSE
// Expected: false, Output: false

console.log(
  `Can attend all appointments: ${can_attend_all_appointments2([
    new Interval(6, 7),
    new Interval(2, 4),
    new Interval(8, 12),
  ])}`
); // Expected: true, Output: true

console.log(
  `Can attend all appointments: ${can_attend_all_appointments2([
    new Interval(4, 5),
    new Interval(2, 3),
    new Interval(3, 6),
  ])}`
); // Expected: false, Output: false

File: assignment_solutions/mergeIntervals/lc252/lc252.py
# Problem Statement #

# Given an array of intervals representing ‘N’ appointments, find out if a person can attend all the appointments.

def can_attend_all_appointments(intervals):
  # TODO: Write your code here
  return False


def main():
  print("Can attend all appointments: " + str(can_attend_all_appointments([[1, 4], [2, 5], [7, 9]])))
  print("Can attend all appointments: " + str(can_attend_all_appointments([[6, 7], [2, 4], [8, 12]])))
  print("Can attend all appointments: " + str(can_attend_all_appointments([[4, 5], [2, 3], [3, 6]])))


main()




# Solution
# -----
# def can_attend_all_appointments(intervals):
#   intervals.sort(key=lambda x: x[0])
#   start, end = 0, 1
#   for i in range(1, len(intervals)):
#     if intervals[i][start] < intervals[i-1][end]:
#       # please note the comparison above, it is "<" and not "<="
#       # while merging we needed "<=" comparison, as we will be merging the two
#       # intervals having condition "intervals[i][start] == intervals[i - 1][end]" but
#       # such intervals don't represent conflicting appointments as one starts right
#       # after the other
#       return False
#   return True

#   -----

# Time complexity #
# The time complexity of the above algorithm is O(N*logN), where ‘N’ is the total number of appointments. Though we are iterating the intervals only once, our algorithm will take O(N * logN) since we need to sort them in the beginning.

# Space complexity #
# The space complexity of the above algorithm will be O(N), which we need for sorting. For Java, Arrays.sort() uses Timsort, which needs O(N) space.
